src/v/security/oidc_service.cc

```diff
@@ -19,67 +23,113 @@
 #include <seastar/core/lowres_clock.hh>
 #include <seastar/core/sstring.hh>
 #include <seastar/coroutine/as_future.hh>
+#include <seastar/coroutine/exception.hh>
+#include <seastar/util/defer.hh>
 
 #include <absl/algorithm/container.h>
+#include <absl/container/flat_hash_map.h>
+#include <boost/outcome/success_failure.hpp>
+
+#include <chrono>
 
 namespace security::oidc {
 
 namespace {
-ss::future<ss::sstring> make_request(std::string_view url_view) {
-    auto url_res = security::oidc::parse_url(url_view);
-    if (url_res.has_error()) {
-        throw std::runtime_error("Invalid url");
+
+template<typename... Args>
+[[nodiscard]] auto return_exception(
+  std::error_code ec,
+  fmt::format_string<Args...> fmt,
+  Args&&... args) noexcept {
+    return ss::coroutine::return_exception(
+      exception(ec, fmt::format(fmt, std::forward<Args>(args)...)));
+}
+
+using seastar::operator co_await;
+
+class probe {
+public:
+    using hist_t = log_hist_public;
+
+    class auto_measure {
+    public:
+        explicit auto_measure(probe* p)
+          : _p{p}
+          , _m{p->_request_latency} {}
+
+    public:
+        void success() noexcept { _p->request_success(std::move(_m)); }
+        void failed() noexcept { _p->request_error(std::move(_m)); }
+
+    private:
+        probe* _p;
+        hist_t::measurement _m;
+    };
```

Comment:
Not trying to be too pedantic here, but should you have some sort of protection to ensure `success()` and `failure()` are not called multiple times? 
```suggestion
    class auto_measure {
    public:
        explicit auto_measure(probe* p)
          : _p{p}
          , _m{p->_request_latency} {}

    public:
        void success() noexcept {
          if (!_reported) {
            _p->request_success(std::move(_m)); 
            _reported = true;
          }
        void failed() noexcept { 
          if (!_reported) {
            _p->request_error(std::move(_m)); 
            _reported = true;
          }
        }

    private:
        bool _reported{false};
        probe* _p;
        hist_t::measurement _m;
    };
```

Reply:
Or have the user set a flag and do the remove in the dtor?

Reply:
It's a fair point..

Reply:
Re-engineered `probe`.